File: Python/7-Trees/Algorithms/1-Depth-FirstSearch(DFS)/Depth-FirstSearch(DFS).py
from typing import Optional, List
from collections import deque
import threading
import asyncio

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        """
        Initializes a TreeNode object.

        Args:
            val (int): The value of the node. Defaults to 0.
            left (Optional[TreeNode]): The left child of the node. Defaults to None.
            right (Optional[TreeNode]): The right child of the node. Defaults to None.
        """
        self.val = val
        self.left = left
        self.right = right

# Approach 1: Recursive DFS (Preorder)
def dfs_recursive_preorder(root: Optional[TreeNode]) -> List[int]:
    """
    Performs a Depth-First Search (DFS) traversal of a binary tree using recursion.
    This implementation uses a preorder traversal strategy:
    1. Visit the current node.
    2. Recursively traverse the left subtree.
    3. Recursively traverse the right subtree.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[int]: A list containing the node values in preorder traversal order.
    """
    result = []  # Initialize an empty list to store the traversal result

    def dfs(node: Optional[TreeNode]):
        """
        Recursive helper function to perform the DFS traversal.

        Args:
            node (Optional[TreeNode]): The current node being visited.
        """
        if not node:
            return  # Base case: if the node is None, return (stop recursion)
        result.append(node.val)  # Preorder: Visit the node *before* its children
        dfs(node.left)          # Recursively traverse the left subtree
        dfs(node.right)         # Recursively traverse the right subtree

    dfs(root)  # Start the DFS traversal from the root of the tree
    return result  # Return the list of node values in preorder order

# Approach 2: Iterative DFS Using Stack (Preorder)
def dfs_iterative_preorder(root: Optional[TreeNode]) -> List[int]:
    """
    Performs a Depth-First Search (DFS) traversal of a binary tree using a stack.
    This implementation uses a preorder traversal strategy:
    1. Visit the current node.
    2. Push the right child onto the stack (if it exists).
    3. Push the left child onto the stack (if it exists).
    The left child is pushed last so it will be processed first (LIFO).

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[int]: A list containing the node values in preorder traversal order.
    """
    if not root:
        return []  # Base case: if the root is None, return an empty list
    result = []  # Initialize an empty list for the result
    stack = [root]  # Initialize a stack with the root node

    while stack:
        node = stack.pop()          # Pop the top node from the stack
        result.append(node.val)    # Visit the node (preorder)

        if node.right:
            stack.append(node.right)  # Push the right child onto the stack *first*
        if node.left:
            stack.append(node.left)   # Push the left child onto the stack *last*
            #  (so it's processed *next*)
    return result  # Return the preorder traversal list

# Approach 3: Multi-threaded DFS (Preorder)
def dfs_multithreaded(root: Optional[TreeNode]) -> List[int]:
    """
    Attempts to perform a Depth-First Search (DFS) traversal of a binary tree using
    multiple threads.  This implementation uses a preorder traversal strategy.
    However, due to Python's Global Interpreter Lock (GIL), true parallelism is not
    achieved, and this approach is generally *not* recommended for CPU-bound tasks
    like tree traversal.  It's included here for demonstration and comparison.  A lock
    is used to protect the shared result list.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[int]: A list containing the node values in preorder traversal order.
    """
    if not root:
        return []  # Base case: if the root is None, return an empty list
    result = []  # Initialize the result list
    lock = threading.Lock()  # Initialize a lock to protect access to 'result'

    def dfs(node: Optional[TreeNode]):
        """
        Recursive helper function to perform DFS in a thread.

        Args:
            node (Optional[TreeNode]): The current node being visited.
        """
        if not node:
            return  # Base case: if the node is None, return
        with lock:
            result.append(node.val)  # Preorder: Visit the node before its children.
            #  Use a lock to ensure thread-safe access to the shared 'result' list.
        left_thread = threading.Thread(target=dfs, args=(node.left,))  # Create thread for left subtree
        right_thread = threading.Thread(target=dfs, args=(node.right,)) # Create thread for right subtree
        left_thread.start()         # Start the left subtree thread
        right_thread.start()        # Start the right subtree thread
        left_thread.join()          # Wait for the left subtree thread to finish
        right_thread.join()         # Wait for the right subtree thread to finish

    dfs(root)  # Start the DFS traversal from the root
    return result  # Return the preorder traversal list

# Approach 4: Async DFS Using Coroutine (Preorder-like)
async def dfs_async(root: Optional[TreeNode]) -> List[int]:
    """
    Attempts to perform a Depth-First Search (DFS) traversal of a binary tree using
    asynchronous operations with `asyncio`.  This implementation is *not* a standard
    preorder traversal.  The use of a queue and `asyncio.sleep` makes it behave
    more like a level-order traversal with artificial delays, and the order of
    node processing is not guaranteed to be strictly DFS.  It's included as an
    example of using `asyncio` with a tree, but it's not a correct DFS implementation.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        List[int]: A list containing the node values, but *not* in standard preorder order.
    """
    if not root:
        return []
    result = []
    queue = deque([root])

    async def process_node(node: Optional[TreeNode]):
        """
        Asynchronous helper function to process a node.
        """
        if not node:
            return
        result.append(node.val)
        await asyncio.sleep(0.01)  # Simulate an asynchronous operation (e.g., I/O)
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)

    while queue:
        node = queue.popleft()
        await process_node(node)

    return result

# Approach 5: DFS Using Generator (Preorder)
def dfs_generator(root: Optional[TreeNode]):
    """
    Performs a Depth-First Search (DFS) traversal of a binary tree using a generator.
    This implementation uses a preorder traversal strategy and yields the node
    values one at a time.  Generators are memory-efficient for potentially large trees.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Yields:
        int: The value of the next node in the preorder traversal.
    """
    if not root:
        return  # Base case: if the root is None, return (generator stops)
    yield root.val        # Preorder: Yield the node's value *before* its children
    yield from dfs_generator(root.left)  # Yield values from the left subtree
    yield from dfs_generator(root.right) # Yield values from the right subtree

# Sample Tree
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))

# Execute different DFS methods
print("Recursive Preorder DFS:", dfs_recursive_preorder(tree))
print("Iterative Preorder DFS:", dfs_iterative_preorder(tree))
print("Multithreaded DFS:", dfs_multithreaded(tree))
print("Generator-based DFS:", list(dfs_generator(tree))) #convert generator to list for printing

# Run Async DFS
async def run_async():
    result = await dfs_async(tree)
    print("Async DFS:", result)

asyncio.run(run_async())

File: Java/7-Trees/Algorithms/1-Depth-FirstSearch(DFS)/Depth-FirstSearch(DFS).java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class DFSExample {

    public static void main(String[] args) {
        // Create a sample binary tree
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);

        System.out.println("DFS Traversal (Recursive - Preorder):");
        dfsRecursivePreorder(root); // Output: 1 2 4 5 3 6 7
        System.out.println();

        System.out.println("DFS Traversal (Recursive - Inorder):");
        dfsRecursiveInorder(root); // Output: 4 2 5 1 6 3 7
        System.out.println();

        System.out.println("DFS Traversal (Recursive - Postorder):");
        dfsRecursivePostorder(root); // Output: 4 2 5 6 3 7 1
        System.out.println();

        System.out.println("DFS Traversal (Iterative - Stack):");
        dfsIterativeStack(root); // Output: 1 2 4 5 3 6 7 (Preorder equivalent)
        System.out.println();

        System.out.println("DFS Traversal (Iterative - Stack - Postorder):");
        dfsIterativeStackPostorder(root);
        System.out.println();

        System.out.println("DFS Traversal (Iterative - Stack - Inorder):");
        dfsIterativeStackInorder(root);
        System.out.println();
    }

    // 1. Recursive Depth-First Search (Preorder Traversal: Root -> Left -> Right)
    //   - Time Complexity: O(N), where N is the number of nodes in the tree.
    //   - Space Complexity: O(H), where H is the height of the tree (due to the call stack).  In the worst case (skewed tree), H can be N, so it can be O(N).  In a balanced tree, it's O(log N).
    public static void dfsRecursivePreorder(TreeNode node) {
        if (node == null) {
            return; // Base case: if the node is null, we've reached the end of a branch, so we stop.
        }

        System.out.print(node.val + " "); // Process the current node (Root).
        dfsRecursivePreorder(node.left);  // Recursively traverse the left subtree.
        dfsRecursivePreorder(node.right); // Recursively traverse the right subtree.
    }

    // 2. Recursive Depth-First Search (Inorder Traversal: Left -> Root -> Right)
    public static void dfsRecursiveInorder(TreeNode node) {
        if (node == null) {
            return;
        }
        dfsRecursiveInorder(node.left);
        System.out.print(node.val + " ");
        dfsRecursiveInorder(node.right);
    }

    // 3. Recursive Depth-First Search (Postorder Traversal: Left -> Right -> Root)
    public static void dfsRecursivePostorder(TreeNode node) {
        if (node == null) {
            return;
        }
        dfsRecursivePostorder(node.left);
        dfsRecursivePostorder(node.right);
        System.out.print(node.val + " ");
    }



    // 4. Iterative Depth-First Search (Preorder Traversal using a Stack)
    //   - Time Complexity: O(N), where N is the number of nodes.
    //   - Space Complexity: O(H), where H is the height of the tree.  In the worst case, O(N) for a skewed tree.
    public static void dfsIterativeStack(TreeNode root) {
        if (root == null) {
            return;
        }

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root); // Start by pushing the root onto the stack.

        while (!stack.isEmpty()) {
            TreeNode node = stack.pop(); // Get the top node from the stack.
            System.out.print(node.val + " "); // Process the node.

            // Important: Push the right child first, so the left child is processed next.
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
            //This ensures that the left child is processed before the right child, maintaining the Preorder traversal order.
        }
    }

    // 5. Iterative Depth-First Search (Postorder Traversal using two Stacks)
    public static void dfsIterativeStackPostorder(TreeNode root) {
        if (root == null)
            return;

        Stack<TreeNode> stack1 = new Stack<>();
        Stack<TreeNode> stack2 = new Stack<>();

        stack1.push(root);

        while (!stack1.isEmpty()) {
            TreeNode temp = stack1.pop();
            stack2.push(temp);

            if (temp.left != null)
                stack1.push(temp.left);
            if (temp.right != null)
                stack1.push(temp.right);
        }

        while (!stack2.isEmpty()) {
            System.out.print(stack2.pop().val + " ");
        }
    }

    // 6. Iterative Depth-First Search (Inorder Traversal using Stack)
    public static void dfsIterativeStackInorder(TreeNode root) {
        if (root == null) {
            return;
        }

        Stack<TreeNode> stack = new Stack<>();
        TreeNode curr = root;

        while (curr != null || !stack.isEmpty()) {
            // Keep going left until you find a null
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }

            // Process the top of the stack
            curr = stack.pop();
            System.out.print(curr.val + " ");

            // Go right
            curr = curr.right;
        }
    }
}

// Definition of a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
    }
}